src/test/ui-fulldeps/fluent-messages/test.stderr

```diff
@@ -41,5 +41,25 @@ help: previously defined in this resource
 LL |         a => "./duplicate-a.ftl",
    |         ^
 
-error: aborting due to 4 previous errors
+error: name `this-slug-has-hyphens` contains a '-' character
+  --> $DIR/test.rs:63:9
+   |
+LL |         slug_with_hyphen => "./slug-with-hyphen.ftl",
+   |         ^^^^^^^^^^^^^^^^
+   |
+   = help: replace any '-'s with '_'s
+
+error: proc macro panicked
```

Comment:
Do you think you could have the `proc_macro` still work in this case? Regardless of whether the user uses a casing that we will accept in the end, we can generate the correct output and then the rest of the compiler can continue and report more errors to the user.

Reply:
I don't think proc macros support non-fatal "emit but continue until the very last moment" style errors (it's still an unstable feature, maybe it could be suggested?). The closest would probably be a warning, which probably (haven't tried) would in combination with the `-Dwarnings` flag passed to all rustc crated be the same as a non-fatal error. But even that won't be  what you ask for, because that also stops compilation at the errors crate.

However, there is a way: adding it as a tidy lint. This however would mean that tidy requires the capability to parse the ftl files. I'm not sure how heavyweight the fluent parsing crate is, need to research that.

Generally it seems that there is no error for referencing invalid internal slugs (those starting with a -), instead it just doesn't emit those slugs. So it might be an idea to implement a linter to check for those and maybe other stuff like slugs that never get used etc.

As one of the other improvements I've been thinking about: All of rustc depending on this error crate is not optimal, as a change of one of the messages might mean a recompilation of the entire chain. Instead, it would be better to have each crate only depend on the ftl file it uses, so to move the invocation of the `fluent_messages` macro into the single crates. Ideally this would also mean that unused lints work now. For panics of the proc macro this would mean that the error is more localized.

Reply:
> I don't think proc macros support non-fatal "emit but continue until the very last moment" style errors (it's still an unstable feature, maybe it could be suggested?). The closest would probably be a warning, which probably (haven't tried) would in combination with the -Dwarnings flag passed to all rustc crated be the same as a non-fatal error. But even that won't be what you ask for, because that also stops compilation at the errors crate.

I think we should be able to avoid this "proc macro panicked" error though - in all of the other examples from this test, we just produce the error we emit without an additional "proc macro panicked" error. As far as I can tell, it's just because the code we go on to generate has probably looks something like this?

```rust
let "this-slug-has-hyphens": DiagnosticMessage = /* .. */;
```

> As one of the other improvements I've been thinking about: All of rustc depending on this error crate is not optimal, as a change of one of the messages might mean a recompilation of the entire chain. Instead, it would be better to have each crate only depend on the ftl file it uses, so to move the invocation of the fluent_messages macro into the single crates. Ideally this would also mean that unused lints work now. For panics of the proc macro this would mean that the error is more localized.

I'm also okay with this. Moving the generation of the typed identifiers into each crate should be relatively straightforward, but the harder part will be constructing the fallback `FluentBundle` in such a way that one crate doesn't depend on every `ftl` file (that's why we do it this way right now). If you manage it, we could ditch the crate prefixes from slugs if we did that and just by-convention have a `rustc_crate_name::fluent` module containing the diagnostic messages. I'm pretty sure we don't have any slugs depending on slugs from other files right now.

Reply:
> I think we should be able to avoid this "proc macro panicked" error though - in all of the other examples from this test, we just produce the error we emit without an additional "proc macro panicked" error. As far as I can tell, it's just because the code we go on to generate probably looks something like this?

I've changed it back to replace `-` with `_` but still emit an error. I hope this should be good for now? We can talk about moving all the errors into tidy at a later point.

Reply:
I'm skeptical of the necessity of doing any checks in `tidy`, your changes seem to have removed the "proc macro panicked" error that was unnecessary.